src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNIPacket.cs

```diff
@@ -247,37 +247,50 @@ public void Reset()
         /// <param name="callback">Completion callback</param>
         public void ReadFromStreamAsync(Stream stream, SNIAsyncCallback callback)
         {
-            bool error = false;
-            
-            stream.ReadAsync(_data, 0, _capacity, CancellationToken.None).ContinueWith(t =>
+            // Treat local function as a static and pass all params otherwise as async will allocate
+            async Task ReadFromStreamAsync(SNIPacket packet, SNIAsyncCallback cb, ValueTask<int> valueTask)
             {
-                Exception e = t.Exception?.InnerException;
-                if (e != null)
+                bool error = false;
+                try
                 {
-                    SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, SNICommon.InternalExceptionError, e);
-                    error = true;
-                }
-                else
-                {
-                    _length = t.Result;
-
-                    if (_length == 0)
+                    packet._length = await valueTask.ConfigureAwait(false);
+                    if (packet._length == 0)
                     {
                         SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, 0, SNICommon.ConnTerminatedError, string.Empty);
                         error = true;
                     }
                 }
+                catch (Exception ex)
+                {
+                    SNILoadHandle.SingletonInstance.LastError = new SNIError(SNIProviders.TCP_PROV, SNICommon.InternalExceptionError, ex);
+                    error = true;
+                }
 
                 if (error)
                 {
-                    Release();
+                    packet.Release();
                 }
 
-                callback(this, error ? TdsEnums.SNI_ERROR : TdsEnums.SNI_SUCCESS);
-            },
-            CancellationToken.None,
-            TaskContinuationOptions.DenyChildAttach,
-            TaskScheduler.Default);
+                cb(packet, error ? TdsEnums.SNI_ERROR : TdsEnums.SNI_SUCCESS);
+            }
+
+            ValueTask<int> vt = stream.ReadAsync(new Memory<byte>(_data, 0, _capacity), CancellationToken.None);
```

Comment:
Can you verify if my understanding of the intended behavior is correct?

The stream.ReadAsync posts a network read and if it successfully completes (synchronously) then the callback is invoked because `vt.IsCompletedSuccessfully` returned true. 
`vt.IsCompletedSuccessfully == false` means that network read is still pending or could have errored out and the client should wait for the valueTask to allocated a Task for completion.

Is this what you were intending? 
 
Can an error check be done on `vt`, before calling `ReadFromStreamAsync` to not allocate in case the stream read errors out? Or is it only possible to get the error by awaiting the ValueTask? 

IIUC, the optimization here is that if `stream.ReadAsync` completes synchronously, then we could be preventing Task allocation in those paths.

Reply:
> Is this what you were intending?

Yes, though it should still allocate less than ContinueWith for the error path.

> Can an error check be done on vt, before calling ReadFromStreamAsync to not allocate in case the stream read errors out?

Yes, though would need to read `.Result` inside a try/catch, catching the exception to then add to the callback, so be doing extra logic and exception handling for the sync completion.

Is an error a common path vs successful completion? Is a Task much extra allocation on top of an exception with stacktrace? With exception handling being a petty slow path anyway is it worth optimizing the sync completion for it?

Reply:
That's a fair argument. Now I understand why you did what you did 😃